injector/src/main/java/edu/ucr/cs/riple/injector/modifications/Deletion.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2022 Nima Karimipour
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package edu.ucr.cs.riple.injector.modifications;
+
+import com.github.javaparser.Position;
+import java.util.List;
+import javax.lang.model.element.ElementKind;
+
+/** Represents a deletion of a content in the source file. */
+public class Deletion extends Modification {
+
+  /** The end position of content which should be removed. */
+  private final Position endPosition;
+
+  public Deletion(String content, Position startPosition, Position endPosition, ElementKind kind) {
+    super(content, startPosition, kind);
+    this.endPosition = new Position(endPosition.line - 1, endPosition.column - 1);
+  }
+
+  @Override
+  public void visit(List<String> lines) {
+    int start = startPosition.column;
+    int end = endPosition.column;
+    StringBuilder line = new StringBuilder(lines.get(startPosition.line));
+    line.delete(start, end + 1);
+    // char at start is removed, head is one character before start.
+    int head = start - 1;
+    // if head is surrounded by white space, remove the extra whitespace.
+    // (  ...) -> ( ...)
+    if (head > 1 && line.charAt(head - 1) == ' ' && line.charAt(head) == ' ') {
+      line.deleteCharAt(head);
+    }
+    // if head is not alphabetic and is followed by a whitespace, remove it.
```

Comment:
Yes,  Actually just the check for non-alphanumerical is enough for all cases. The reason for this check it is that we would like to remove the space inserted by additions on deletions. Please see the example below:
```java
void foo(Object bar)
```
If we make `bar` `@Nullable`, we get:
```java
void foo(@Nullable Object bar)
```

If we just delete the annotation later we will leave the empty space inserted by addition and produce output below:
```java
void foo( Object bar)
```

Please note we cannot always delete the space in between, consider the example below:
```java
public@annot Object foo();
```
If always delete the space we will mix two tokens and produce `publicObject foo();`, therefore we should only remove the space if there exists a non-alphanumerical to keep tokens separated.

Added a test case as well: 7bfb20a65d0e9ebd3b48800e464ed0059f1021c7

Reply:
> Yes, Actually just the check for non-alphanumerical is enough for all cases. 

I don't think that's correct. Consider this horrifying but JLS-valid annotation name:

```
@interface AnnotationWithNonAlphanumericSuffix$$$$_ { }

class Foo {
    public void foo(@AnnotationWithNonAlphanumericSuffix$$$$_@Nullable Object o) { }
}
```

I suspect removing `@Nullable` there will give you broken code. The right check is probably `Character.isJavaIdentifierPart(...)`, rather than `Character.isLetterOrDigit(...)`... but the simpler and most robust thing to do would probably be just `Character.isWhitespace(...)` and drop the negation.

Reply:
Hi @lazaroclapp, thank you for catching this bug, yes the current implementation fails with the test case you provided. I changed it to using `isJavaIdentifierPart(...)` and it worked along all other test cases.

`Character.isWhitespace(...)` does not work as it does not remove the left white space in addition. Please see the example below for test case `edu.ucr.cs.riple.injector.RemovalTest > remove_annot_param`:
For `@Nullable Object test(@Nullable Object o)` 
we get:   `@Nullable Object test( Object o)`
and not: `@Nullable Object test(Object o)`

Please let me know if I am missing something. Thank you. 


Also added a test case that only passes if we use `Character.isJavaIdentifierPart(...)`. eed362882eade8a4046e7530237ed1e859308237

Reply:
@msridhar @lazaroclapp Actually I don't think we should rely on GJF too much and should do the fixing on our end as much as possible. The reason is that during the process of measuring effects of fix trees (at least for now) we operate multiple pairs of `@Nullable` addition / deletion on one element at each depth. If our consecutive addition / deletion leave a space, in the end wight end up with a method like below:

```java
void foo(                 Object f,                   Object c)
```

Although this is fixable by GJF (which we do not run it within iterations) this can be the end state of source files at iteration `j` and the input for iteration `j+1`. To make sure we will not run the parser on complicated cases within iterations, it is better to not rely on GJF and fix it ourself.